lms/djangoapps/bookmarks/tests/test_views.py

```diff
@@ -0,0 +1,462 @@
+            3) With empty request.DATA
+        """
+        # Send usage_id with invalid format.
+        response = self.send_post(
+            client=self.client,
+            url=reverse('bookmarks'),
+            json_data={'usage_id': 'invalid'},
+            expected_status=400
+        )
+        self.assertEqual(response.data['user_message'], "Invalid usage id")
+        self.assertEqual(response.data['developer_message'], "Invalid usage id")
+
+        # Send data without usage_id.
+        response = self.send_post(
+            client=self.client,
+            url=reverse('bookmarks'),
+            json_data={'course_id': 'invalid'},
+            expected_status=400
+        )
+        self.assertEqual(response.data['user_message'], "No usage id provided")
+        self.assertEqual(response.data['developer_message'], "No usage id provided")
+
+        # Send empty data dictionary.
+        response = self.send_post(
+            client=self.client,
+            url=reverse('bookmarks'),
+            json_data={},
+            expected_status=400
+        )
+        self.assertEqual(response.data['user_message'], "No data provided")
+        self.assertEqual(response.data['developer_message'], "No data provided")
+
+    def test_post_bookmark_for_non_existing_block(self):
+        """
+        Test that posting a bookmark for a block that does not exist returns a 400.
+        """
+        response = self.send_post(
+            client=self.client,
+            url=reverse('bookmarks'),
+            json_data={'usage_id': 'i4x://arbi/100/html/340ef1771a094090ad260ec940d04a21'},
+            expected_status=400
+        )
+        self.assertEqual(response.data['user_message'], "Invalid usage id")
+        self.assertEqual(response.data['developer_message'], "Block with usage_id not found.")
+
+    def test_unsupported_methods(self):
+        """
+        Test that DELETE and PUT are not supported.
+        """
+        self.client.login(username=self.user.username, password=self.test_password)
+        self.assertEqual(405, self.client.put(reverse('bookmarks')).status_code)
+        self.assertEqual(405, self.client.delete(reverse('bookmarks')).status_code)
+
+
+class BookmarksDetailViewTests(BookmarksViewTestsMixin):
+    """
+    This contains the tests for GET & DELETE methods of bookmark.views.BookmarksDetailView class
+    """
+    def test_get_bookmark_successfully(self):
+        """
+        Test that requesting bookmark returns data with 200 code.
+        """
+        response = self.send_get(
+            client=self.client,
+            url=reverse(
+                'bookmarks_detail',
+                kwargs={'username': self.user.username, 'usage_id': unicode(self.vertical_1.location)}
+            )
+        )
+        data = response.data
+        self.assertIsNotNone(data)
+        self.assert_valid_bookmark_response(data, self.bookmark_1)
+
+    def test_get_bookmark_with_optional_fields(self):
```

Comment:
Nit: you could use DDT to provide the value of optional_fields and collapse this test and the previous into just one test that is run twice.